Migrate train schedules to paced trains

flomonster opened the pull request:
closes #14263 
closes #14264
closes #14265

The idea of this PR is to merge train schedules and paced trains together.

- The application should work the same
- All train schedules are migrated to paced trains
- All train schedules endpoints become useless (will be clean later)
- Endpoints working with paced trains should be adapted:
  - Conflict & requirements: Use a new unify `TrainId`
  - Simulation, Pathfinding, Projection... : Handle the case of a "single" paced train
- New export and import format


### TODO

- [x] Make migration
  - Check it works whith existing paced trains ( ✔️ )
- [x] Adapt schema and models
- [x] Adapt conflict and requirements
- [x] Adapt Simulation, PF, Projection, Occupancy, Track Occupancy
  - No need, all these endpoints handle the base train schedule
- [x] Disable train schedule creation
- [x] Adapt frontend
- [x] Adapt integration tests
- [x] Adapt e2e tests
- [x] Check that everything works

flomonster committed:
editoast: add migration to move train schedules to paced trains

- Make paced fields nullable
- Move train schedules to paced trains
- Move round trips with them

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

clarani committed:
front: make paced nullable in TrainSpaceTimeData

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: transform Train type to handle single PacedTrain

The Train type can now be a TrainScheduleId, an OccurrenceId or a PacedTrainId.
After the end of the migration, TrainScheduleId won't exist anymore (see the coming PRs).

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: do not create trainSchedules anymore, but pacedTrains instead

- ManageTimetableItem view
  - when creating a train from ManageTimetableItem view
  - when turning a paced train into a train schedule
- NGE: when changing the frequence of a train
- ImportTimetableItem view: when importing a trainSchedule
- Trains list: when duplicating a train

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

flomonster committed:
editoast, front: make paced field optional

editoast
- Adapt models and schemas
- Rework TrainId to better handle the merge of train schedule and paced
  trains

front
- Introduce temporary types of PacedTrains with paced field defined: PacedTrainWithPaced, PacedTrainResponseWithPaced
- Handle a paced train without paced field as a TrainSchedule (single train)
- Fix the conflicts/occupations endpoints and data workflow
- Fix labels when editing a pacedTrain without paced
- Fix RoundTrips suggestions for PacedTrains without paced
- Fix selected items labels to handle PacedTrains without paced
- Fix selectedTrainId process when creating or editing pacedTrains without paced field:
In useAutoSelectTrainId, if the train to select is a pacedTrain without paced, then the id to
select is its raw id, and not the id of its first occurrence. It's the same when editing a train.

The naming may not be ideal, but all this code will be cleaned in the coming weeks.

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>
Co-authored-by: Clara Ni <clara.ni@outlook.fr>

flomonster commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +324,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
+        id: i64,
+        exception_key: String,
     },
 }
```
This need to be discussed. I unified `OccurrenceId` and `TrainId`.
I opted for three variants, but I could also go for four like:

```rust
pub enum TrainId {
    Single {
        id: i64,
    },
    BaseOccurrence {
        id: i64,
        index: u64,
    },
    ModifiedOccurrence {
        id: i64,
        index: u64.
        exception_key: String,
    },
    CreatedOccurrence {
        id: i64,
        exception_key: String,
    },
Comment view}
```

woshilapin replied:
I would go for 4 variants, like it was done before, i.e., the choice to separate a “regular occurrence” from a “modified occurrence”. I find it clearer to have one `enum` with 4 variants, than an `enum` with 3 variants, for which one of them have a sub-variant (`Option`) which is semantically less clear (`Option` can mean anything, it’s a generic type, whereas `BaseOccurrence`[^1] and `ModifiedOccurrence` are distinguishable in our business jargon).

[^1]:  or just `Occurrence` , I’m fine with a shorter name

woshilapin replied:
The more I read this PR, the more I’m convinced that we should just represent what you call “single” as just an occurrence and therefore, have a `TrainId` like this.

```rust
pub enum Occurrence {
    Base {
        id: i64,
        index: u64,
    },
    Modified {
        id: i64,
        index: u64,
        exception_key: String,
    },
    Created {
        id: i64,
        exception_key: String,
    }
}
```

Note that in this case, there is still the `paced` field that is optional, which enforce the fact that there cannot be any exception applied on a `Base` in the case of a single train. Encoding like this will simplify the `get_base_occurrences` and `num_base_occurrences` (no special `if` is needed anymore). Let’s decide on that before the rest.

flomonster replied:
I'm not sure to understand your suggestion. You'll like to represent single train like `Occurrence::Base{id: ts_id, index: 0}` ?

flomonster replied:
I fixed this comment and the following as discussed.

I reworked the "TrainId" enum to make it an "Occurrence" enum. Things are more natural with the new enum. Let me know if the new version seems fine for you.

woshilapin replied:
Good for me (just a small comment on the naming). I’ll let you resolve since you initiate the thread.

woshilapin commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +324,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
+            exception_key: None,
+        }
+    }
+}
```
Not sure that we need those since we directly construct the others (`CreatedOccurrence`, see Line 143, or `Occurrence` Line 173) from the `enum`. At least, let’s create constructors for either all of them, or none of them (slight favor of none in this case since it’s straightforward, but either way, I’ll approve).

clarani committed:
front: drop dangerous function extractEditoastIdFromTrainId

extractEditoastIdFromTrainId is dangerous because it loses information about
where the ID came from (TrainSchedule, PacedTrain...).
Moreover, the function is not used, so it can be removed easily.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

woshilapin commented on editoast/src/models/paced_train.rs:
It feels like we’re still trying to handle 2 different objects (“train schedule” vs “paced train”) inside the same object, and it does create some inconsistencies. For example, I would expect that `get_base_occurrences` returns a `Vec` of length `num_base_occurrences`, but that is not the case for `Single`.

Technically, `get_base_occurrences` makes only sense as a temporary state to apply exceptions. Therefore, it’s only useful for “occurrences”, not “single”. If we still want to keep the distinction between “single” and ”occurrences”, I would advise to have `get_base_occurrences` return an empty `Vec` for “single”.

woshilapin replied:
The new version makes this comment outdated. I’m resolving.

woshilapin commented on editoast/src/views/pagination.rs:
Why is `ConcatenatedPaginatedList` removed? Doesn’t seem related with the current PR at least, so maybe in a different PR?

flomonster replied:
I adapted conflict and requirements endpoints which don't need to retrieve train schedule anymore. It was the only place where `ConcatenatedPaginatedList` was used so I removed it (since I needed an annotation to dismiss warnings).

clarani committed:
front: refacto getEditLabel

Since the PacedTrain without a paced field will soon replace the TrainSchedules,
we won't be able to distinguish single trains from paced trains only with
the id (it will always be PacedTrains).
We will then have to check the paced field of the timetableItem to do it.

To prepare this migration, getEditLabel now receives the timetableItem instead of simply
the timetableItemId.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: allow import paced trains without paced

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

flomonster committed:
editoast: remove train schedule creation endpoint

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

woshilapin commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +313,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
+pub enum Occurrence {
+    Base {
+        id: i64,
+        index: u64,
```
I’m wondering why we do not use `usize` here since indexes in Rust are usually `usize`. It would avoid some `as` conversion in some places.

woshilapin commented on editoast/src/models/paced_train.rs:
```diff
@@ -245,10 +265,19 @@ impl From<paced_train::PacedTrain> for PacedTrainChangeset {
-            .exceptions(paced.exceptions);
+            .options(train_schedule_base.options);
+
+        changeset = if let Some(paced) = paced {
```
nitpick: Not necessary to do an `if` here. It makes the code more linear to follow (feel free to agree/disagree and resolve).

```rust
.time_window(paced.map(|p| p.time_window).map(ChronoDuration::from))
.interval(paced.map(|p| p.interval).map(ChronoDuration::from))
.exceptions(paced.map(|p| p.exceptions).unwrap_or_default()
```

flomonster replied:
The borrow checker don't want your change :cry:  I'll try something to improve this part of the code.

woshilapin replied:
Yep, I forgot, you need to take references (`.as_ref()`)… and it will work once you add `Copy` trait to `PositiveDuration` (it’s fine since it wraps a type that is already `Copy`).

```rust
            .time_window(
                paced
                    .as_ref()
                    .map(|p| p.time_window)
                    .map(ChronoDuration::from),
            )
            .interval(paced.as_ref().map(|p| p.interval).map(ChronoDuration::from))
            .exceptions(paced.map(|p| p.exceptions).unwrap_or_default())

```

woshilapin commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +313,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
-    PacedTrain {
-        paced_train_id: i64,
-        occurrence_id: OccurrenceId,
+pub enum Occurrence {
```
I wouldn’t say this is an occurrence, since it doesn’t contain any schedule information. How about naming it `OccurrenceId`?

woshilapin commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +313,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
+        Occurrence::Base { id, index }
+    }
+
+    /// Create a new `Occurrence::Modified` without an exception key.
```
Why “without an exception key”?

clarani committed:
front: refacto DeleteModal

Gives selected elements count instead of giving the arrays of selected ids, since
we don't need the arrays of ids.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev commented on front/src/common/api/generatedEditoastApi.ts:
```diff
@@ -2140,23 +2129,24 @@ export type PostPacedTrainTrackOccupancyApiResponse =
     [key: string]: ((
       | {
+          id: number;
           index: number;
-          type: 'BaseOccurrence';
+          type: 'base';
         }
       | {
           exception_key: string;
+          id: number;
           index: number;
-          type: 'ModifiedException';
+          type: 'modified';
         }
       | {
           exception_key: string;
-          type: 'CreatedException';
+          id: number;
+          type: 'created';
         }
```
nit: This part is repeated twice (also in the `Conflict` type), it could be nice to have this extracted in its own type

clarani replied:
Can we leave this for a future PR ? :pray:

SharglutDev replied:
sure !

SharglutDev commented on front/src/utils/trainId.ts:
```diff
@@ -206,15 +206,6 @@ export const extractPacedTrainIdFromOccurrenceId = (occurrenceId: OccurrenceId):
- * Given a train id with a TrainId format (either TrainScheduleId or OccurrenceId),
- * returns the Editoast id (used for api).
- */
-export const extractEditoastIdFromTrainId = (id: TrainId): number =>
```
Why is it "dangerous" ? The function is unused, maybe the commit description should rather say that

clarani replied:
It's dangerous because if we use this function, we don't know anymore if we are manipulating a paced Train or a TrainSchedule. We should not keep a generic function (and it's not used)

clarani committed:
front: generalize TrainScheduleItem props

Instead of giving a TrainSchedule, the component TrainScheduleItem can now receive
a TimetableItem.
In the coming commits, the paced train without paced field will be handled by this
component, as a TrainSchedule (single train).

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

Caracol3 commented on front/src/utils/trainId.ts:
Shouldn't we had `isPacedTrainId` here too now ?

clarani replied:
You right, thanks !!

woshilapin commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +313,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
-}
-
-#[derive(Debug, Clone, Hash, PartialEq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Hash)]
```
We probably don’t want `ToSchema` in an object that is part of `models` if we want to eventually move this to `editoast_models`. We might need another `views` object like it was the case before (there was `models::TrainId` and `views::PacedTrainOccurrenceId`).

flomonster replied:
I prefer keep it this way and duplicate the struct only when we move this to models.

Caracol3 commented on front/src/applications/operationalStudies/hooks/useEtcsBrakingCurves.ts:
```diff
@@ -61,9 +64,9 @@ const useEtcsBrakingCurves = (
       return undefined;
-    if (isTrainScheduleId(selectedTrainId))
+    if (!isOccurrenceId(selectedTrainId))
       throw new Error(`trainId ${selectedTrainId} should be a occurrence id`);
```
If i understand correctly, the selected train can now be the paced train base item and not only an occurence. 
Doesn't this means that here, the UI can crash even if this is a valid selection ?

clarani replied:
Yes exactly, since we will replace the TrainSchedules by PacedTrains without paced fields :) 
Concerning the UI, I have not modified the existing behaviour, I just added some code to handle a pacedTrain without paced field like a TrainSchedule. Normally, the UI will not crash (I tested and there are also the e2e tests) !

Caracol3 reviewed (commented):
Nice work !
Just left two questions.

Caracol3 reviewed (approved):
LGTM (tested)

SharglutDev commented on front/src/reducers/osrdconf/types.ts:
```diff
@@ -224,11 +224,17 @@ export type TrainScheduleWithTrainId = TrainSchedule & {
 export type TimetableItem = TrainScheduleWithTrainId | PacedTrainWithPacedTrainId;
-export type Train = TrainScheduleWithTrainId | TrainBaseWithOccurrenceId;
+export type Train =
+  | TrainScheduleWithTrainId
```
Maybe we should rename this one by `TrainBaseWithTrainScheduleId` ?

clarani replied:
Can we do this later ?

SharglutDev replied:
Sure

SharglutDev commented on front/src/common/BootstrapSNCF/ModalSNCF/DeleteModal.tsx:
```diff
@@ -1,38 +1,37 @@
-  selectedPacedTrainIds: PacedTrainId[];
-  selectedTrainScheduleIds: TrainScheduleId[];
+  selectedPacedTrainsCount?: number;
+  selectedTrainSchedulesCount?: number;
```
```suggestion
  selectedPacedTrainCount?: number;
  selectedTrainScheduleCount?: number;
```

clarani replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -36,7 +36,7 @@ const useScenarioData = (scenario: ScenarioWithDetails, infraId: number) => {
   const [timetableItems, setTimetableItems] = useState<TimetableItem[]>();
   const timetableItemsById = useMemo(() => mapBy(timetableItems, 'id'), [timetableItems]);
-  const [selectedTimetableItemIds, setSelectedTimetableItemIds] = useState<TimetableItemId[]>([]);
+  const [selectedTimetableItems, setSelectedTimetableItems] = useState<TimetableItem[]>([]);
```
Not sure to understand why we need to keep the whole timetable item in this state ?

clarani replied:
Because with only the timetableItemId, we are not able to separate the trainSchedules from the pacedTrains anymore (a PacedTrain without a paced field should be considered as a trainSchedule). I can add some details in the commit

SharglutDev replied:
Yes please :)

clarani replied:
It's done !

SharglutDev commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -22,12 +22,17 @@ import type {
 
 export type Board = 'trains' | 'map' | 'macro' | 'std' | 'sdd' | 'tables' | 'conflicts';
 
+export type PacedTrainWithPaced = TrainSchedule & { paced: NonNullable<PacedTrain['paced']> };
```
Shouldn't this type just be name `PacedTrain` ? A paced train will always have `paced` otherwise it's a trainschedule

clarani replied:
The type `PacedTrain` already exists :/ It's only a temporary type

SharglutDev replied:
Yeah sorry for the comment, I'm a bit 🤯 after the review 😅

SharglutDev commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -22,12 +22,17 @@ import type {
 export type Board = 'trains' | 'map' | 'macro' | 'std' | 'sdd' | 'tables' | 'conflicts';
 
+export type PacedTrainWithPaced = TrainSchedule & { paced: NonNullable<PacedTrain['paced']> };
+export type PacedTrainResponseWithPaced = PacedTrainWithPaced & {
+  id: PacedTrainId;
+};
```
It's confusing to have this type and `PacedTrainWithPacedTrainId`.
We should use paced train only when `paced` is define.

clarani replied:
I would rather keep this naming till we drop the `PacedTrainId`. I agree with you that it is not a `PacedTrain` technically (since it has no paced field), but it has a `PacedTrainId`

flomonster committed:
editoast, front: search: adapt train schedule search with the migration and use it for linkedTrains

Searching "trainschedule" will look in the paced train table.
The workflow for linked trains will now search in the paced train table.

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>
Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ManageTimetableItem/helpers/formatTimetableItemPayload.ts:
```diff
@@ -116,10 +120,16 @@ export function formatPacedTrainPayload(
+  if (
+    timetableItemToEditData &&
+    isPacedTrainToEditData(timetableItemToEditData) &&
+    timetableItemToEditData.originalPacedTrain.paced
```
Shouldn't `paced` in `PacedTrainWithDetails` be required ?

clarani replied:
No because you can have a PacedTrain without paced :/

SharglutDev commented on front/src/modules/conflict/utils.ts:
```diff
@@ -13,73 +13,63 @@ function getConflictTrainNames(
-      return computeOccurrenceName(pacedTrain.train_name, occurrence.index);
+  const trainNames: string[] = [];
+  conflict.train_ids.forEach((train) => {
+    const pacedTrain = trainMap.get(formatEditoastIdToPacedTrainId(train.id));
```
Maybe it's just me but it's pretty confusing to call a `TimetableItem` a `pacedTrain` could be call it just `train` for example ?

clarani replied:
At this step, we know that we have a `PacedTrain` (as its id will be a `PacedTrainId`). I have the feeling that there is a confusion coming from the naming here:
- `PacedTrain` are trains object with a special type of id (`PacedTrainId`), but it can either be a paced train or a single train, depending on whether they have a paced field defined or not
- `PacedTrainWithPaced` corresponds to a real paced train, with a paced field defined

This naming convention may be a bit confusing, but it's only temporary and improving it would require much work and a longer review.

SharglutDev commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -108,6 +108,14 @@ export const formatPacedTrainWithDetails = (
+      ...extractBaseTimetableItemProps(pacedTrain),
+      rollingStock,
+      summary: formatSummary(pacedTrain, pacedTrainSummary?.paced_train),
+    };
```
nit: could we keep using {} for `if`s that don't fit in one line plz :)

SharglutDev commented on front/src/modules/timetableItem/helpers/pacedTrain.ts:
```diff
@@ -30,10 +36,15 @@ export const isPacedTrainResponseWithPaced = (
 ): timetableItem is PacedTrainResponseWithPaced =>
   isPacedTrainResponseWithPacedTrainId(timetableItem) && !!timetableItem.paced;
 
+export const isPacedTrainWithPacedWithDetails = (
```
We need to find another way to name this 🫤

clarani replied:
This function is only temporary ;) The PR to remove it is already drafted

SharglutDev commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/helpers/configureHandlePan.ts:
```diff
@@ -91,7 +91,7 @@ export function configureHandlePan({
           ({ id }) => isPacedTrainId(id) && id === pacedTrainId
         );
-        if (pacedTrain && 'paced' in pacedTrain) {
+        if (pacedTrain && !isTrainScheduleProjection(pacedTrain) && pacedTrain.paced) {
```
In all previous commits, we try to get rid of "TrainSchedule" utils but here we add one ?

clarani replied:
Yes, because we don't have a function isPacedTrainProjection, and since it will be removed with the TrainSchedule endpoints/types/utils in the next PR, I didn't want to create a new utils

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/CalendarTrainList.tsx:
```diff
@@ -151,7 +151,13 @@ const CalendarTrainList = ({
+              isSelected={
+                workerStatus === 'READY' &&
+                selectedTrainId ===
+                  (isPacedTrainWithDetails(timetableItem)
```
Pretty confusing 🤔

clarani replied:
It's only for a week or 2, the time of the migration !

SharglutDev reviewed (commented):
Thank you for the PR !

For what it's worth, I'm really not a fan of this new naming convention. I'm afraid it will be a source of confusion while the rest of the merge feature is in process.

> [!WARNING]
> It seems I can't duplicate a train schedule anymore (but I built a while ago so maybe one of the last changes you did fixed it)

aschantraine commented on editoast/src/views/search.rs:
The test `search_trainschedule_post_found()` needs to be adapted.

flomonster replied:
Fixed

aschantraine reviewed (approved):
One test needs to be adapted, but otherwise it seems to work well.

clarani committed:
front: simplify logic about train updates

In the previous commits, the existing logic was kept:  when updating a
paced train without a paced field (i.e in trainSchedule mode), the existing
paced train was deleted and a new one was created (as if we were transforming
a paced train into a new train schedule, because the existing item has a PacedTrainId
and is considered as a paced train).

This workflow broke end-to-end tests.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

Maymanaf commented on front/tests/utils/setup-utils.ts:
```diff
@@ -194,7 +193,7 @@ export async function createDataForTests(): Promise<void> {
       )
     ).scenario;
-    await sendTrainSchedules(scenarioWithTimetableItems.timetable_id, trainSchedulesJson);
+    await sendPacedTrains(scenarioWithTimetableItems.timetable_id, trainSchedulesJson);
```
It might be clearer to rename it to `sendTrains`

Maymanaf reviewed (approved):
LGTM for e2e tests ✅

emersion commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -35,9 +37,9 @@ const useSimulationResults = (): SimulationResults | undefined => {
+  const train: Train | undefined = useMemo(() => {
     if (!selectedTrainId || !timetableItem) return undefined;
-    if (!isPacedTrainResponseWithPacedTrainId(timetableItem)) {
+    if (!isOccurrenceId(selectedTrainId) || !isPacedTrainResponseWithPacedTrainId(timetableItem)) {
```
Instead of adding a new condition here, shouldn't we change the `isTrainScheduleId()` condition below?

Adding a new condition here is a behavior change and makes the check below dead code.

clarani replied:
Hum if I modify only the condition in the check below, I won't correctly handle pacedTrains without paced (they should be handled like TrainSchedules). Indeed it's a behaviour change.

I thought it was the right moment to introduce this change, but maybe not. Do you want me to move it in a following commit ?

emersion replied:
Before this commit ("front: transform Train type to handle single PacedTrain"), only two cases were handled: train schedules and occurrences. This commit introduces a third case, paced trains, which should be handled exactly like train schedules. That is, the only thing this commit needs to do is mechanically replace `isTrainScheduleId()` checks with `!isOccurrenceId()`.

Paced trains without paced can be handled by checking for `timetableItem.paced` in addition to `isPacedTrainResponseWithPacedTrainId()` in this condition in one of the subsequent commits.

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -72,7 +74,12 @@ const useSimulationResults = (): SimulationResults | undefined => {
-    if (!selectedTrainId || !timetableItem || !isPacedTrainResponseWithPacedTrainId(timetableItem))
+    if (
+      !selectedTrainId ||
+      !isOccurrenceId(selectedTrainId) ||
```
Ditto

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/CalendarTrainList.tsx:
```diff
@@ -148,11 +144,11 @@ const CalendarTrainList = ({
             <TrainScheduleItem
               isInSelection={selectedTimetableItemIds.includes(timetableItem.id)}
               handleSelectTrain={handleSelectTimetableItem}
-              train={timetableItem as TrainScheduleWithDetails}
```
Oh nice, we can get rid of this cast :)

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/TrainScheduleItem.tsx:
```diff
@@ -83,6 +84,7 @@ const TrainScheduleItem = ({
   };
 
   const deleteTrain = async () => {
+    if (isPacedTrainWithDetails(train)) return;
```
Hm, this sounds dangerous. If this is temporary, can we at least add a big fat TODO to drop this?

emersion replied:
Since a future PR in the same commit drops it, let's just ignore this.

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/TrainScheduleItem.tsx:
```diff
@@ -103,6 +105,8 @@ const TrainScheduleItem = ({
   };
 
   const duplicateTrain = async () => {
+    if (isPacedTrainWithDetails(train)) return;
```
Ditto

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/TrainScheduleItem.tsx:
```diff
@@ -41,14 +42,14 @@ import { formatTrainDuration, getTrainCategoryClassName } from './utils';
 type TrainScheduleItemProps = {
   isInSelection: boolean;
-  train: TrainScheduleWithDetails;
+  train: TimetableItemWithDetails;
```
I'm not sure I understand this change. At the end of this PR, this piece of code should only handle paced trains. Why change it to a type which can be either? Is the intention to first accept both train schedules and paced trains here, then correctly handle paced trains (instead of early returns), then drop the train schedule case?

clarani replied:
Yes exactly. At the end of this PR, we won't be able to create `TrainSchedules` anymore, but the rest of the code will still be able to handle both `TrainSchedules` and `PacedTrains`. The drop of `TrainSchedules` will be done in another PR

emersion replied:
Alright!

emersion commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -36,7 +36,7 @@ const useScenarioData = (scenario: ScenarioWithDetails, infraId: number) => {
   const [timetableItems, setTimetableItems] = useState<TimetableItem[]>();
   const timetableItemsById = useMemo(() => mapBy(timetableItems, 'id'), [timetableItems]);
-  const [selectedTimetableItemIds, setSelectedTimetableItemIds] = useState<TimetableItemId[]>([]);
+  const [selectedTimetableItems, setSelectedTimetableItems] = useState<TimetableItem[]>([]);
```
I don't think storing the full selected timetable item is a good idea here. The stored data will be outdated as soon as the original item is mutated. If a paced train is selected, then turned into a one-time train by editing it (via NGE or the OSRD modal), `selectedTimetableItems` will not be updated. (This is true for other train properties as well.)

Can we instead keep only the ID in `selectedTimetableItems`, and grab the full timetable item from `timetableItemsById` when we need it?

clarani replied:
Ok, I can drop this commit :+1:

emersion commented:
Note, `Co-authored-by` can be used to indicate when a patch has been authored by more than 1 person (for instance, " editoast, front: search: adapt train schedule search with the migration and use it for linkedTrains").

emersion commented on editoast/src/views/search.rs:
```diff
@@ -736,41 +736,41 @@ pub(super) struct SearchResultItemScenario {
 #[cfg_attr(test, derive(serde::Deserialize))]
 #[search(
-    table = "train_schedule",
+    table = "paced_train",
```
Don't we need to generate a migration here?

flomonster replied:
No need to generate anything, this is only used to build the search query (I tested it 😉)

emersion replied:
Alright!

emersion commented on front/src/common/api/generatedEditoastApi.ts:
This change breaks the front build. This commit is not atomic.

emersion commented on editoast/src/models/paced_train.rs:
```diff
@@ -284,53 +309,69 @@ impl From<PacedTrain> for paced_train::PacedTrain {
-    BaseOccurrence { index: u64 },
-    ModifiedException { index: u64, exception_key: String },
-    CreatedException { exception_key: String },
+    Base {
```
It would've been nicer to keep this rename in a separate commit, but too late I guess.

emersion reviewed (commented):
I've reviewed up to commit "editoast: make paced field optional". Will continue soon.

emersion commented on front/src/applications/operationalStudies/types.ts:
```diff
@@ -22,12 +22,17 @@ import type {
 
 export type Board = 'trains' | 'map' | 'macro' | 'std' | 'sdd' | 'tables' | 'conflicts';
 
+export type PacedTrainWithPaced = TrainSchedule & { paced: NonNullable<PacedTrain['paced']> };
```
Shouldn't the base type be `PacedTrain` instead of `TrainSchedule` here?

Could we perhaps add a TODO to rename this to `PacedTrain` once the migration is complete?

clarani replied:
I already have a draft PR for the renaming

emersion replied:
Turns out `TrainSchedule` contains just base fields without any `id`.

emersion commented on front/src/modules/timetableItem/helpers/pacedTrain.ts:
```diff
@@ -18,6 +22,14 @@ import {
 
 import type { ExceptionChangeGroups, PacedTrainWithDetails } from '../types';
 
+export const isPacedTrainWithPaced = (pacedTrain: PacedTrain): pacedTrain is PacedTrainWithPaced =>
```
Do we really need this helper? Can't we just use `if (pacedTrain.paced)` instead of introducing this function?

clarani replied:
I need it for typescript checks :/ But it's only temporary

emersion replied:
Which case is TypeScript unhappy about?

emersion commented on front/src/modules/simulationResult/components/SpaceTimeChartWrapper/useTrackOccupancy.ts:
```diff
@@ -238,40 +238,56 @@ const useTrackOccupancy = ({
+            if (!exception) throw new Error(`Exception not found for train ${train.id}`);
+
+            if (occupation.type === 'created') {
+              if (!exception.start_time?.value) continue;
```
A created exception necessarily has a start time. Since a bunch of invalid states were turned into errors above, should we also throw an exception here?

clarani replied:
Done

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/ManageTimetableItem/helpers/formatTimetableItemPayload.ts:
```diff
@@ -116,10 +120,16 @@ export function formatPacedTrainPayload(
     const originalPacedTrain = formatPacedTrainWithDetailsToPacedTrainPayload(
       timetableItemToEditData.originalPacedTrain
     );
+    if (!isPacedTrainWithPaced(originalPacedTrain)) return newPacedTrain;
```
This should never happen right? We've checked above that `timetableItemToEditData.originalPacedTrain.paced` is non-null. Should this be turned into an error?

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/ManageTimetableItem/helpers/__tests__/formatTimetableItemPayload.spec.ts:
```diff
@@ -185,7 +186,7 @@ describe('formatTimetableItemPayload', () => {
         },
       },
-    },
+    } as PacedTrainWithPacedWithDetails,
```
Could we define the variable type instead of using a cast here, as was done before?

(Same below)

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/SimulationResults/SimulationResults.tsx:
```diff
@@ -142,6 +142,9 @@ const SimulationResults = ({
     );
     if (!pacedTrain || !isPacedTrainWithDetails(pacedTrain)) return undefined;
+
+    if (!pacedTrain.paced) return pacedTrain.summary;
```
This is never supposed to happen because `selectedTrainId` is an occurrence ID, I think?

In that case, we should probably throw an error here?

clarani replied:
You're right, done

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/Timetable/TrainScheduleItem.tsx:
```diff
@@ -105,51 +107,46 @@ const TrainScheduleItem = ({
-    })
-      .unwrap()
-      .catch((e) => {
-        dispatch(setFailure(castErrorToFailure(e)));
```
We're loosing the error handling here. Can we wrap the `getPacedTrain()` call in a try/catch?

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/ManageTimetableItem/hooks/useUpdateTimetableItem.ts:
```diff
@@ -91,9 +91,10 @@ const useUpdateTimetableItem = (
+        (selectedTrainId &&
+          isOccurrenceId(selectedTrainId) &&
+          extractPacedTrainIdFromOccurrenceId(selectedTrainId) === timetableItemId) ||
+        !updatedItem.paced
```
Shouldn't this be an error? Since `simulationConf.editingItemType !== 'trainSchedule'`, shouldn't `updatedItem.paced` always be non-null?

In that case, this should be turned into an error.

emersion replied:
Ah, this goes away in commit "front: simplify logic about train updates". Probably not worth bothering with then.

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTimetableItem/helpers/postTimetableItems.ts:
```diff
@@ -1,44 +1,28 @@
     }));
   }
-  return { trainSchedules, pacedTrains };
+  return { timetableItems };
```
Can't we return `timetableItems` directly here without wrapping the list in an object?

clarani replied:
Done

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTimetableItem/ImportTimetableItemTrainsList.tsx:
```diff
@@ -175,20 +166,10 @@ const ImportTimetableItemTrainsList = ({
-        formattedTrainSchedules,
-        extractEditoastIdFromTrainScheduleId
-      );
-      requests.push(postTrainScheduleRoundTrips(payload).unwrap());
```
We have no use for `requests` anymore.

clarani replied:
Done

emersion reviewed (commented):
Thanks a lot for splitting this PR into commits!

I think the fix commits at the end of the PR should be squashed into earlier commits. A commit should not leave the app in a bad state, even if TypeScript doesn't complain the changes in earlier commits make the app misbehave. For instance "front: fix labels when editing a pacedTrain without paced" should be squashed into "front: make paced field optional in PacedTrainWithDetails".

We haven't really talked about how this interacts with paced trains without an absolute start time. This is important because when we designed paced trains, the main difference with train schedules was the eventual relative start time. If we merge train schedules and paced trains, then this change becomes trickier. Would it look something like this?

```ts
type TrainSchedule = {
  train_name: string;
  // other common fields
} & (
  | { paced?: { start_offset: string; duration: string; /* etc */ } }
  | { one_time?: { start_time: string } }
);
```

Or, would all trains be missing an absolute start time, and one-time trains would be paced trains with a zero duration? Or something else?

flomonster commented:
> We haven't really talked about how this interacts with paced trains without an absolute start time. This is important because when we designed paced trains, the main difference with train schedules was the eventual relative start time. If we merge train schedules and paced trains, then this change becomes trickier. Would it look something like this?

This question feels out of scope since we did not designed paced train without an absolute start time. We supposed that it was possible only for paced trains but it might not be the case. IMO we should not care about it for now. And design relative start time in afterwards.

emersion reviewed (approved):
Thanks! Feel free to resolve the last remaining comment after squashing the relevant commits!

SharglutDev reviewed (approved):
Lgtm !
